apollo-router/src/spec/query.rs

```diff
@@ -640,21 +624,18 @@ impl Query {
                     }
 
                     if name.as_str() == TYPENAME {
-                        let input_value = input
-                            .get(field_name.as_str())
-                            .cloned()
-                            .filter(|v| v.is_string())
-                            .unwrap_or_else(|| {
-                                Value::String(ByteString::from(
-                                    current_type.inner_named_type().as_str().to_owned(),
-                                ))
```

Comment:
here we tried to use __typename values returned by subgraph
but we don't validate this value at all it can even be a name of interface if subgraph uses `@interfaceObject`

By GraphQL spec `__typename` should always use object name (interfaces and unions are forbidden), so I change code to first use `current_type` (always valid) and fallback to using subgraph's value only if `current_type` is not an object type.

Reply:
this should be checked with what the output rewriting code in the QP is doing

Reply:
Not sure what to do here.
Realistically, I can't inspect QP in the scope of this PR.
I can revert this particular change, because it separated from other fixes.
@Geal Do you think it worth to revert it?

Reply:
The change looks correct to me: when a field is an object type in the supergraph we know what the correct `__typename` is and could even skip asking the subgraph